reader_concurrency_semaphore.cc

```diff
@@ -438,6 +438,14 @@ class reader_permit::impl
         }
     }
 
+    const tracing::trace_state_ptr& trace_state() const noexcept {
+        return _trace_ptr;
+    }
+
+    void set_trace_state(tracing::trace_state_ptr trace_ptr) noexcept {
+        _trace_ptr = std::move(trace_ptr);
```

Comment:
(continuing from below). What if _trace_ptr is set?

Perhaps we should create a stitching record that says that old_session continues in new_session.

Reply:
Note that files etc will all have the old session in them, if they captured it.

Reply:
I'll respond to the page-bridging question here: if I set the trace pointer on the first page and leave it as-is, all reader concurrency semaphore messages will be in the first session. In the example from the cover letter, all messages would be in the tracing session "8cc80d50-c72d-11ed-8427-14e21cc3ed56".
If I refresh the trace pointer on each page, then each message goes into the session of the relevant page. We seems to have one session per page.

Reply:
Yes, but let's have an event "resuming read from session {}" or something, to create a link between pages (and to note the fact that we have an 1865 hit).

Reply:
@avikivity each trace_state is defined by a corresponding `session_id` value, which in line corresponds to a single CQL query.
A single CQL (read) query is always limited to a single page.
Hence each new page has to have a new trace_state object if traced.

And now the most interesting question is when that "if traced" holds.

When tracing is enabled from the application (will eventually come down to Scylla using a Trace bit) then things are rather straight forward exactly as @denesb shown in the commit message example - every new page will show up as new session object. And I agree with you, it would definitely make sense to create some way to "connect" different pages. And hence what you suggested (create a trace point that would have "old session id -> new session id" info) makes a lot of sense.

However, just as a thought, when Tracing is enable via probabilistic tracing we may trace some pages out of multi-page read sequence and not the others.

In such a case, on the one hand, it would be interesting to trace the whole multi-page sequence. But on the other hand, implementing this would require us to change the current semantics and the implementation of the probabilistic tracing (quite) a bit.

I don't have a strong opinion but from the debugging perspective having the whole multi-page sequence traced if the first page was picked would make a lot of sense.

Doesn't have to be part of this series however - just a thought.

Reply:
> and to note the fact that we have an 1865 hit

We already have a trace point for this:
```
                                                                                                                               Reusing querier [shard 0] | 2023-03-20 10:43:17.395614 | 127.0.0.1 |            116 | 127.0.0.1
 ```

Reply:
Ok, let's add a link to the previous session then.

Reply:
I was about to say that this requires intrusive changes like saving the session id in the paging state and whatnot... then remembered that in `set_trace_state()` I have the old trace state and the new one and I can just use the new one to create a trace point with the session id extracted from the old one. I hope it will be as simple as it sounds written down.

Reply:
Guess I'll fix this in a follow-up.

Reply:
Follow-up here: https://github.com/scylladb/scylladb/pull/13318.